File: src/menus/providers/skolmaten/fetch.ts
import { URLSearchParams } from "url";
import urljoin from "url-join";
import { fetchRetry } from "../../../utils/fetch-retry";

export interface SkolmatenRequestOptions {
	path: string;
}

export function getSkolmatenUrl(path: string, searchParams?: URLSearchParams): string {
	let queryString = searchParams?.toString() ?? "";

	if (queryString.length > 0) {
		queryString = `?${queryString}`;
	}

	return urljoin("https://skolmaten.se/api/4", path, queryString);
}

export default async function skolmatenFetch<T>(path: string, searchParams?: URLSearchParams): Promise<T> {
	const url = getSkolmatenUrl(path, searchParams);

	const res = await fetchRetry(url, {
		headers: {
			"API-Version": "4.0",
			"Client-Token": "web",
			"Client-Version-Token": "web",
			Locale: "sv_SE",
		},
	});

	const data: T = await res.json();

	return data;
}

File: src/menus/providers/skolmaten/fetch.rs
use reqwest::Client;
use serde::de::DeserializeOwned;

pub async fn fetch_json<T: DeserializeOwned>(client: &Client, path: &str) -> reqwest::Result<T> {
    let url = format!("https://skolmaten.se/api/4/{}", path);

    let res = client
        .get(&url)
        .header("API-Version", "4.0")
        .header("Client-Token", "web")
        .header("Client-Version-Token", "web")
        .header("Locale", "sv_SE")
        .send()
        .await?;

    let body = res.json::<T>().await?;

    Ok(body)
}
